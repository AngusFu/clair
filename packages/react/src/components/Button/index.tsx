import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const types = [
  "primary",
  "default",
  "warning",
  "danger",
  "success",
  "ghost"
] as const;
export type Type = typeof types[number];
const PType = PropTypes.oneOf([...types]);

const sizes = ["large", "normal", "small"] as const;
export type Size = typeof sizes[number];
const PSize = PropTypes.oneOf([...sizes]);

const htmlTypes = ["button", "reset", "submit"] as const;
export type HtmlType = typeof htmlTypes[number];
const PHtmlType = PropTypes.oneOf([...htmlTypes]);

// handle "detail" property of custom event
type ButtonEvent = React.MouseEvent<HTMLButtonElement> &
  Partial<Pick<CustomEvent, "detail">>;
type EventHandler = React.EventHandler<ButtonEvent>;

export interface ButtonProps {
  size?: Size;
  type?: Type;
  htmlType?: HtmlType;
  disabled?: boolean;
  loading?: boolean;
  block?: boolean;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
  onClick?: EventHandler;
  onSubmit?: EventHandler;
  onReset?: EventHandler;
  forwardRef?: React.RefObject<HTMLButtonElement>;
}

export const Button: React.FC<ButtonProps> = props => {
  const {
    size = "normal",
    type = "default",
    htmlType = "button",
    disabled = false,
    loading = false,
    block = false,
    children,
    forwardRef,
    className,
    ...rest
  } = props;

  const className2Use: string = classNames("c-button", className, {
    [`c-button--${type}`]: types.some(t => t === type),
    [`c-button--${size}`]: sizes.some(s => s === size),
    "c-button--block": block,
    "c-button--loading": loading
  });
  return (
    <button
      ref={forwardRef}
      className={className2Use}
      disabled={disabled}
      type={htmlType}
      {...rest}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  size: PSize,
  type: PType,
  htmlType: PHtmlType,
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  loading: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func,
  onSubmit: PropTypes.func,
  onReset: PropTypes.func,
  children: PropTypes.node,
  forwardRef: PropTypes.shape({
    current: PropTypes.any
  })
} as PropTypes.ValidationMap<ButtonProps>;

Button.displayName = "Button";
