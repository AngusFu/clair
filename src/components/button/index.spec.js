import { shallowMount, mount } from '@vue/test-utils'
import Button from './index.vue'

it('should works with different themes and flavors', () => {
  const snapshotTest = propsData => {
    const wrapper = shallowMount(Button, {
      propsData,
      slots: { default: 'My Button' }
    })
    expect(wrapper).toMatchSnapshot()
  }

  const themes = ['default', 'primary', 'success', 'warning', 'danger']
  const flavors = ['round', 'outline', 'flat']

  for (let theme of themes) {
    for (let flavor of flavors) {
      const propsData = {}
      if (theme !== 'default') {
        propsData[theme] = true
      }
      propsData[flavor] = true
      snapshotTest(propsData)
    }
  }
})

it('should works with feather icon', () => {
  const wrapper = mount(Button, {
    propsData: { icon: 'send' },
    slots: { default: 'My Button' }
  })
  expect(wrapper.find('svg')).toMatchSnapshot()
  expect(wrapper).toMatchSnapshot()
})

it('should emit event when clicked', () => {
  let clicked = false
  const wrapper = mount(Button, {
    slots: { default: 'My Button' },
    listeners: {
      click () {
        clicked = true
      }
    }
  })
  wrapper.trigger('click')
  expect(clicked).toBe(true)
  expect(wrapper).toMatchSnapshot()
})

it('could be disabled', () => {
  let clicked = false
  const wrapper = mount(Button, {
    propsData: { disabled: true },
    slots: { default: 'My Button' },
    listeners: {
      click () {
        clicked = true
      }
    }
  })

  wrapper.trigger('click')
  expect(clicked).toBe(false)

  expect(wrapper.find('button').attributes('disabled')).toBe('disabled')
  expect(wrapper).toMatchSnapshot()
})
