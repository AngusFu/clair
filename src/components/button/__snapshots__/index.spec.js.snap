// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`could be disabled 1`] = `
<button type="button" class="c-button" disabled="disabled">
  <!----><span>My Button</span></button>
`;

exports[`should emit event when clicked 1`] = `
<button type="button" class="c-button">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 1`] = `
<button type="button" class="c-button c-button--round">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 2`] = `
<button type="button" class="c-button c-button--outline">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 3`] = `
<button type="button" class="c-button c-button--flat">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 4`] = `
<button type="button" class="c-button c-button--primary c-button--round">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 5`] = `
<button type="button" class="c-button c-button--primary c-button--outline">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 6`] = `
<button type="button" class="c-button c-button--primary c-button--flat">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 7`] = `
<button type="button" class="c-button c-button--success c-button--round">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 8`] = `
<button type="button" class="c-button c-button--success c-button--outline">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 9`] = `
<button type="button" class="c-button c-button--success c-button--flat">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 10`] = `
<button type="button" class="c-button c-button--warning c-button--round">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 11`] = `
<button type="button" class="c-button c-button--warning c-button--outline">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 12`] = `
<button type="button" class="c-button c-button--warning c-button--flat">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 13`] = `
<button type="button" class="c-button c-button--danger c-button--round">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 14`] = `
<button type="button" class="c-button c-button--danger c-button--outline">
  <!----><span>My Button</span></button>
`;

exports[`should works with different themes and flavors 15`] = `
<button type="button" class="c-button c-button--danger c-button--flat">
  <!----><span>My Button</span></button>
`;

exports[`should works with feather icon 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-send" style="vertical-align: middle;">
  <line x1="22" y1="2" x2="11" y2="13"></line>
  <polygon points="22 2 15 22 11 13 2 9 22 2"></polygon>
</svg>
`;

exports[`should works with feather icon 2`] = `<button type="button" class="c-button"><span class="c-icon">â€‹<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-send" style="vertical-align: middle;"><line x1="22" y1="2" x2="11" y2="13"></line><polygon points="22 2 15 22 11 13 2 9 22 2"></polygon></svg></span><span>My Button</span></button>`;
